### PR TITLE
refactor: move migrations to gobuffalo/fizz

### PR DESCRIPTION
**Pull request by aeneasr**
## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
- [ ] I have read the [security policy](../security/policy)
- [ ] I confirm that this pull request does not address a security vulnerability. If this pull request addresses a security
vulnerability, I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation within the code base (if appropriate)
- [ ] I have documented my changes in the [developer guide](https://github.com/ory/docs) (if appropriate)

### CODE REVIEW

**Comment by zepatrik** [comment]
@aeneasr hey this is kinda working now
It currently is in a state where both the old and new migration systems are working so it is easier to test it. You can basically run `hydra migrate sql` and `hydra migrate sqlold`. You can also run `make dumpdbs` to create a (postgres) db dump after applying both the old and new migrations to a fresh database so you can diff them.
What is your recommended way of testing this automatically? We said something like running the quickstart for different versions of hydra and then migrating up but how would we then check whether it worked?
Also I didn't address cockroach migrations yet because they don't use the shared migrations. I think we can add empty files for not existing migrations there and for the next migrations to come use probably the fizz language anyway or trying to use shared sql for cockroach aswell.

**Comment by aeneasr** [comment]
Good job! Let's discuss this on slack/in a hangout. Ping me when you're online :)

**Comment by aeneasr** [comment]
To do:

- [ ] Please dump databases for MySQL and CRDB as well and make a diff to ensure that they are the same
- [ ] CockroachDB support
- [ ] Rewrite migration tests (put them in persistence as a PoC and later merge them into fizz
- [ ] Move all old migrations to `internal/fizzmigrate` as well as all code depending on `sql-migrate` and write tests that run old migrations step by step and then the new migrations afterwards and compare the outputs (the generated sql should be the same always):
  - [ ] Run {1,...,n} old migration and  {1,...,n} new migration (in parallel) and compare their outputs (should be the same)
  - [ ] Run {1,...,n} old migration and  {m-n,...,m} new migrations (in parallel) and compare their outputs (should be the same)

**Comment by zepatrik** [comment]
seems like the docker command is not working in the ci

**Comment by zepatrik** [comment]
done

File: jwk/manager_test.go [hunk 1]
```diff
@@ -40,31 +46,40 @@ var managers = map[string]Manager{}
 var m sync.Mutex
 var testGenerator = &RS256Generator{}
 
-func TestMain(m *testing.M) {
-	runner := dockertest.Register()
-
-	runner.Exit(m.Run())
+func getManager(t *testing.T, url string) Manager {
```
Comment (aeneasr) [inline jwk/manager_test.go:49]: Move to internal

File: jwk/manager_test.go [hunk 2]
```diff
@@ -73,35 +88,16 @@ func TestManager(t *testing.T) {
 	managers["memory"] = reg.KeyManager()
 
 	if !testing.Short() {
-		dockertest.Parallel([]func(){
```
Comment (aeneasr) [inline jwk/manager_test.go:76]: Add parallel boot again

File: oauth2/fosite_store_test.go [hunk 3]
```diff
@@ -103,21 +101,9 @@ func TestManagers(t *testing.T) {
 			registries["memory"] = reg
 
 			if !testing.Short() {
-				var p, m, c *sqlx.DB
```
Comment (aeneasr) [inline oauth2/fosite_store_test.go:106]: Revert to parallel execution

File: oauth2/oauth2_auth_code_test.go [hunk 4]
```diff
@@ -154,32 +152,9 @@ func TestAuthCodeWithDefaultStrategy(t *testing.T) {
 	}
 
 	if !testing.Short() {
-		var p, m, c *sqlx.DB
-		dockertest.Parallel([]func(){
```
Comment (aeneasr) [inline oauth2/oauth2_auth_code_test.go:158]: Revert

File: oauth2/oauth2_refresh_token_test.go [hunk 5]
```diff
@@ -75,24 +72,9 @@ func TestCreateRefreshTokenSessionStress(t *testing.T) {
 		},
 	}
 
-	config := internal.NewConfigurationWithDefaults()
```
Comment (aeneasr) [inline oauth2/oauth2_refresh_token_test.go:78]: revert

File: persistence/sql/migratest/helpers.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/gobuffalo/pop/v5"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
+)
+
```
Comment (aeneasr) [inline persistence/sql/migratest/helpers.go:15]: This is copied from somewhere, link to it!

File: persistence/sql/migratest/helpers.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+		if err := tx.RawQuery(string(data)).Exec(); err != nil {
+			log.Print(mf.Version)
+			return errors.WithStack(err)
+		}
+		return nil
+	}
+
+	if fi, err := os.Stat(migrationPath); err != nil || !fi.IsDir() {
+		// directory doesn't exist
```
Comment (aeneasr) [inline persistence/sql/migratest/helpers.go:86]: Log this message :)

File: persistence/sql/migratest/helpers.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+		t.FailNow()
+		return nil
+	}
+	if fi, err := os.Stat(testDataPath); err != nil || !fi.IsDir() {
+		// directory doesn't exist
```
Comment (aeneasr) [inline persistence/sql/migratest/helpers.go:91]: Log this message :)

File: persistence/sql/migratest/helpers.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+			*tx = *newTx
+		}
+
+		data, err := ioutil.ReadFile(filepath.Join(testDataPath, fileName))
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		if err := tx.RawQuery(string(data)).Exec(); err != nil {
+			log.Print(mf.Version)
```
Comment (aeneasr) [inline persistence/sql/migratest/helpers.go:79]:
```suggestion
			t.Logf(mf.Version)
```

File: persistence/sql/migratest/migration_test.go [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+	"github.com/ory/x/dbal"
+)
+
+func TestMigrations(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+		return
+	}
+
+	for db, connect := range map[string]func(*testing.T) *pop.Connection{
```
Comment (aeneasr) [inline persistence/sql/migratest/migration_test.go:32]: Start these in parallel using `dockertest.Parallel`

File: persistence/sql/migratest/migration_test.go [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+
+					if elor != nil {
+						alor := &consent.LogoutRequest{}
+						require.NoError(t, dbx.Get(alor, dbx.Rebind("select * from hydra_oauth2_logout_request where challenge = ?"), elor.Challenge))
+						assertEqualLogoutRequests(t, elor, alor)
+					}
+				})
+			}
+
+			// TODO this is very stupid and should be replaced as soon the manager uses pop
```
Comment (aeneasr) [inline persistence/sql/migratest/migration_test.go:130]:
```suggestion
			// TODO https://github.com/ory/hydra/issues/1815
			// this is very stupid and should be replaced as soon the manager uses pop
```

File: persistence/sql/migratest/testdata/20190100000001_testdata.sql [hunk 12]
```diff
@@ -0,0 +1 @@
+INSERT INTO hydra_client (id, client_name, client_secret, redirect_uris, grant_types, response_types, scope, owner, policy_uri, tos_uri, client_uri, logo_uri, contacts, public) VALUES ('client-0001', 'Client 0001', 'secret-0001', 'http://redirect/0001_1', 'grant-0001_1', 'response-0001_1', 'scope-0001', 'owner-0001', 'http://policy/0001', 'http://tos/0001', 'http://client/0001', 'http://logo/0001', 'contact-0001_1', true);
```
Comment (aeneasr) [inline persistence/sql/migratest/testdata/20190100000001_testdata.sql:1]: Note to self: figure out how to review this

File: persistence/sql/persister.go [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+const (
+	clientMigrationTableName  oldTableName = "hydra_client_migration"
+	jwkMigrationTableName     oldTableName = "hydra_jwk_migration"
+	consentMigrationTableName oldTableName = "hydra_oauth2_authentication_consent_migration"
+	oauth2MigrationTableName  oldTableName = "hydra_oauth2_migration"
+)
+
+// this type is copied from sql-migrate to remove the dependency
+type OldMigrationRecord struct {
+	Id        string    `db:"id"`
```
Comment (aeneasr) [inline persistence/sql/persister.go:81]:
```suggestion
	ID        string    `db:"id"`
```

**Comment by aeneasr** [comment]
The demo fails with:

```
hydra-migrate_1  | The following migration is planned:
hydra-migrate_1  | 
postgresd_1      | FATAL:  unrecognized configuration parameter "max_idle_conns"
postgresd_1      | FATAL:  unrecognized configuration parameter "max_conns"
hydra-migrate_1  | Could not get the migration status:
hydra-migrate_1  | pq: unrecognized configuration parameter "max_conns"
hydra-migrate_1  | could not create new transaction
hydra-migrate_1  | github.com/gobuffalo/pop/v5.newTX
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/tx.go:28
hydra-migrate_1  | github.com/gobuffalo/pop/v5.(*dB).Transaction
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/db.go:18
hydra-migrate_1  | github.com/gobuffalo/pop/v5.(*Connection).NewTransaction
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/connection.go:169
hydra-migrate_1  | github.com/gobuffalo/pop/v5.(*Connection).Transaction.func1
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/connection.go:136
hydra-migrate_1  | github.com/gobuffalo/pop/v5.commonDialect.Lock
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/dialect_common.go:30
hydra-migrate_1  | github.com/gobuffalo/pop/v5.(*Connection).Transaction
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/connection.go:134
hydra-migrate_1  | github.com/gobuffalo/pop/v5.CreateSchemaMigrations
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/migrator.go:197
hydra-migrate_1  | github.com/gobuffalo/pop/v5.Migrator.CreateSchemaMigrations
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/migrator.go:214
hydra-migrate_1  | github.com/gobuffalo/pop/v5.Migrator.Status
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/migrator.go:219
hydra-migrate_1  | github.com/ory/hydra/persistence/sql.(*Persister).MigrationStatus
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/persistence/sql/persister.go:40
hydra-migrate_1  | github.com/ory/hydra/cmd/cli.(*MigrateHandler).MigrateSQL
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/cmd/cli/handler_migrate.go:76
hydra-migrate_1  | github.com/spf13/cobra.(*Command).execute
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:842
hydra-migrate_1  | github.com/spf13/cobra.(*Command).ExecuteC
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:943
hydra-migrate_1  | github.com/spf13/cobra.(*Command).Execute
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:883
hydra-migrate_1  | github.com/ory/hydra/cmd.Execute
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/cmd/root.go:62
hydra-migrate_1  | main.main
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/main.go:33
hydra-migrate_1  | runtime.main
hydra-migrate_1  |      /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/proc.go:203
hydra-migrate_1  | runtime.goexit
hydra-migrate_1  |      /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/asm_amd64.s:1373
hydra-migrate_1  | couldn't start a new transaction
hydra-migrate_1  | github.com/gobuffalo/pop/v5.(*Connection).NewTransaction
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/connection.go:171
hydra-migrate_1  | github.com/gobuffalo/pop/v5.(*Connection).Transaction.func1
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/connection.go:136
hydra-migrate_1  | github.com/gobuffalo/pop/v5.commonDialect.Lock
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/dialect_common.go:30
hydra-migrate_1  | github.com/gobuffalo/pop/v5.(*Connection).Transaction
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/connection.go:134
hydra-migrate_1  | github.com/gobuffalo/pop/v5.CreateSchemaMigrations
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/migrator.go:197
hydra-migrate_1  | github.com/gobuffalo/pop/v5.Migrator.CreateSchemaMigrations
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/migrator.go:214
hydra-migrate_1  | github.com/gobuffalo/pop/v5.Migrator.Status
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/zepatrik/pop/v5@v5.0.12-0.20200418094557-850d559d8122/migrator.go:219
hydra-migrate_1  | github.com/ory/hydra/persistence/sql.(*Persister).MigrationStatus
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/persistence/sql/persister.go:40
hydra-migrate_1  | github.com/ory/hydra/cmd/cli.(*MigrateHandler).MigrateSQL
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/cmd/cli/handler_migrate.go:76
hydra-migrate_1  | github.com/spf13/cobra.(*Command).execute
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:842
hydra-migrate_1  | github.com/spf13/cobra.(*Command).ExecuteC
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:943
hydra-migrate_1  | github.com/spf13/cobra.(*Command).Execute
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:883
hydra-migrate_1  | github.com/ory/hydra/cmd.Execute
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/cmd/root.go:62
hydra-migrate_1  | main.main
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/main.go:33
hydra-migrate_1  | runtime.main
hydra-migrate_1  |      /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/proc.go:203
hydra-migrate_1  | runtime.goexit
hydra-migrate_1  |      /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/asm_amd64.s:1373
hydra-migrate_1  | github.com/ory/hydra/persistence/sql.(*Persister).MigrationStatus
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/persistence/sql/persister.go:40
hydra-migrate_1  | github.com/ory/hydra/cmd/cli.(*MigrateHandler).MigrateSQL
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/cmd/cli/handler_migrate.go:76
hydra-migrate_1  | github.com/spf13/cobra.(*Command).execute
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:842
hydra-migrate_1  | github.com/spf13/cobra.(*Command).ExecuteC
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:943
hydra-migrate_1  | github.com/spf13/cobra.(*Command).Execute
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:883
hydra-migrate_1  | github.com/ory/hydra/cmd.Execute
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/cmd/root.go:62
hydra-migrate_1  | main.main
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/main.go:33
hydra-migrate_1  | runtime.main
hydra-migrate_1  |      /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/proc.go:203
hydra-migrate_1  | runtime.goexit
hydra-migrate_1  |      /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/asm_amd64.s:1373
hydra-migrate_1  | github.com/ory/hydra/cmd/cli.(*MigrateHandler).MigrateSQL
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/cmd/cli/handler_migrate.go:77
hydra-migrate_1  | github.com/spf13/cobra.(*Command).execute
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:842
hydra-migrate_1  | github.com/spf13/cobra.(*Command).ExecuteC
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:943
hydra-migrate_1  | github.com/spf13/cobra.(*Command).Execute
hydra-migrate_1  |      /Users/foobar/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:883
hydra-migrate_1  | github.com/ory/hydra/cmd.Execute
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/cmd/root.go:62
hydra-migrate_1  | main.main
hydra-migrate_1  |      /Users/foobar/go/src/github.com/ory/hydra/main.go:33
hydra-migrate_1  | runtime.main
hydra-migrate_1  |      /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/proc.go:203
hydra-migrate_1  | runtime.goexit
hydra-migrate_1  |      /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/asm_amd64.s:1373
```

To run the demo, do:

```
$ make docker
$ docker tag oryd/hydra:latest oryd/hydra:v1.4.5
$ docker tag oryd/hydra:latest oryd/hydra:v1.4.6
$ docker-compose -f quickstart.yml \
    -f quickstart-postgres.yml \
    -f quickstart-tracing.yml \
    up --build
```

**Comment by zepatrik** [comment]
@aeneasr quickstart is now working for me and all tests pass, ready for review

File: Makefile [hunk 14]
```diff
@@ -5,15 +5,20 @@ tools:
 		npm i
 		go install github.com/ory/go-acc github.com/ory/x/tools/listx github.com/go-swagger/go-swagger/cmd/swagger github.com/go-bindata/go-bindata/go-bindata github.com/sqs/goreturns github.com/ory/sdk/swagutil
 
+# Runs full test suite including tests where databases are enabled
+.PHONY: test-legacy-migrations
+test-legacy-migrations:
+		make test-resetdb
+		source scripts/test-env.sh && go test -tags legacy_migration_test -failfast -timeout=20m ./internal/fizzmigrate
```
Comment (aeneasr) [inline Makefile:12]: isn't `make sqlbin` missing here?

File: driver/registry.go [hunk 15]
```diff
@@ -56,6 +60,7 @@ func MustNewRegistry(c configuration.Provider) Registry {
 }
 
 func NewRegistry(c configuration.Provider) (Registry, error) {
+	fmt.Printf("got dsn '%s'\n", c.DSN())
```
Comment (aeneasr) [inline driver/registry.go:63]:
```suggestion
```

File: driver/registry.go [hunk 16]
```diff
@@ -1,9 +1,12 @@
 package driver
 
 import (
+	"fmt"
```
Comment (aeneasr) [inline driver/registry.go:4]:
```suggestion
```

File: go.mod [hunk 17]
```diff
@@ -59,3 +60,5 @@ require (
 	golang.org/x/tools v0.0.0-20200313205530-4303120df7d8
 	gopkg.in/square/go-jose.v2 v2.5.0
 )
+
+replace github.com/gobuffalo/pop/v5 => github.com/zepatrik/pop/v5 v5.0.12-0.20200418094557-850d559d8122
```
[the remark this answers is not in the document]
Comment (aeneasr) [reply go.mod:66]: This is still to do @zepatrik

File: internal/driver.go [hunk 18]
```diff
@@ -78,6 +84,101 @@ func NewRegistrySQL(c *configuration.ViperProvider, db *sqlx.DB) *driver.Registr
 	return r.(*driver.RegistrySQL)
 }
 
+func NewRegistryPop(t *testing.T, url string) *driver.RegistrySQL {
```
Comment (aeneasr) [inline internal/driver.go:87]:
```suggestion
func NewRegistrySQL(t *testing.T, url string) *driver.RegistrySQL {
```

File: oauth2/fosite_store_test.go [hunk 19]
```diff
@@ -23,59 +23,49 @@ package oauth2_test
...
-}
-
-func connectToCRDB(t *testing.T) *sqlx.DB {
-	db, err := dockertest.ConnectToTestCockroachDB()
-	require.NoError(t, err)
-	x.CleanSQL(t, db)
-	return db
+var registries = make(map[string]driver.Registry)
+var cleanRegistries = func(*testing.T) {
+	fmt.Printf("\n\nsetting memory reg\n\n\n")
```
Comment (aeneasr) [inline oauth2/fosite_store_test.go:48]:
```suggestion
```

File: Makefile [hunk 20]
```diff
@@ -27,14 +32,13 @@ test-resetdb:
 		docker rm -f hydra_test_database_mysql || true
 		docker rm -f hydra_test_database_postgres || true
 		docker rm -f hydra_test_database_cockroach || true
-		docker run --rm --name hydra_test_database_mysql -p 3444:3306 -e MYSQL_ROOT_PASSWORD=secret -d mysql:5.7
```
[the remark this answers is not in the document]
Comment (zepatrik) [reply Makefile:30]: mistake I think

File: Makefile (hunk 14 above)
Comment (zepatrik) [reply Makefile:12]: hm right I thought we don't need it as the old migrations should not change but if they change we should `make sqlbin` here

**Pull request merged**